pkg/transaction-builder/pocket_transaction_builder.go

```diff
@@ -0,0 +1,154 @@
+type PocketTransactionBuilder struct {
+	provider provider.Provider
+	signer   signer.Signer
+}
+
+// NewPocketTransactionBuilder returns an instance of PocketTransactionBuilder
+func NewPocketTransactionBuilder(provider provider.Provider, signer signer.Signer) *PocketTransactionBuilder {
+	return &PocketTransactionBuilder{
+		provider: provider,
+		signer:   signer,
+	}
+}
+
+func getOptionalParams(options *TransactionOptions) (string, string, int64) {
+	memo := ""
+	coinDenom := Upokt
+	fee := defaultTXFee
+
+	if options != nil {
+		memo = options.Memo
+
+		if options.CoinDenom != "" {
+			coinDenom = options.CoinDenom
+		}
+
+		if options.Fee != 0 {
+			fee = options.Fee
+		}
+	}
+
+	return memo, string(coinDenom), fee
+}
+
+func (t *PocketTransactionBuilder) validateTransactionRequest(chainID string, txMsg TxMsg) error {
+	if t.provider == nil {
+		return ErrNoProvider
+	}
+
+	if t.signer == nil {
+		return ErrNoSigner
+	}
+
+	if chainID == "" {
+		return ErrNoChainID
+	}
+
+	if txMsg == nil {
+		return ErrNoTxMsg
+	}
+
+	return nil
+}
+
+func (t *PocketTransactionBuilder) signTransaction(chainID, memo, coinDenom string, fee int64, txMsg TxMsg) (string, error) {
+	feeStruct := coreTypes.Coins{
+		coreTypes.Coin{
+			Amount: coreTypes.NewInt(fee),
+			Denom:  coinDenom,
+		},
+	}
+
+	entropy, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		return "", err
+	}
+
+	signBytes, err := auth.StdSignBytes(chainID, entropy.Int64(), feeStruct, txMsg, memo)
+	if err != nil {
+		return "", err
+	}
+
+	signature, err := t.signer.GetKeyManager().SignBytes(signBytes)
+	if err != nil {
+		return "", err
+	}
+
+	publicKey, err := crypto.NewPublicKey(t.signer.GetKeyManager().GetPublicKey())
+	if err != nil {
+		return "", err
+	}
+
+	signatureStruct := authTypes.StdSignature{PublicKey: publicKey, Signature: signature}
+
+	tx := authTypes.NewTx(txMsg, feeStruct, signatureStruct, memo, entropy.Int64())
+
+	txBytes, err := auth.DefaultTxEncoder(app.Codec())(tx, -1)
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(txBytes), nil
+}
+
+// CreateTransaction returns input necessary for doing a transaction
+func (t *PocketTransactionBuilder) CreateTransaction(chainID string, txMsg TxMsg, options *TransactionOptions) (*provider.SendTransactionInput, error) {
+	err := t.validateTransactionRequest(chainID, txMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	memo, coinDenom, fee := getOptionalParams(options)
+
+	signedTX, err := t.signTransaction(chainID, memo, coinDenom, fee, txMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &provider.SendTransactionInput{
+		Address:     t.signer.GetKeyManager().GetAddress(),
+		RawHexBytes: signedTX,
+	}, nil
+}
+
+// Submit does the transaction from raw input
+func (t *PocketTransactionBuilder) Submit(chainID string, txMsg TxMsg, options *TransactionOptions) (*provider.SendTransactionOutput, error) {
```

Comment:
Realized that we should probably include sane defaults for `chainID` and some of the transaction options:
- for the `fee` you probably wanna always pay the base fee in uPOKT, 10k.
- For chainID, we should probably default to mainnet and have an enum handy (see above).

Reply:
All optional params have default values setted on `getOptionalParams`

Reply:
Ah cool!